pkg/controllers/management/aks/aks_cluster_handler.go

```diff
@@ -0,0 +1,420 @@
+		}
+		logrus.Infof("waiting for cluster AKS [%s] pre-create failure to be resolved", cluster.Name)
+		return e.SetFalse(cluster, apimgmtv3.ClusterConditionProvisioned, failureMessage)
+	}
+}
+
+func (e *aksOperatorController) setInitialUpstreamSpec(cluster *mgmtv3.Cluster) (*mgmtv3.Cluster, error) {
+	logrus.Infof("setting initial upstreamSpec on cluster [%s]", cluster.Name)
+	upstreamSpec, err := clusterupstreamrefresher.BuildAKSUpstreamSpec(e.SecretsCache, cluster)
+	if err != nil {
+		return cluster, err
+	}
+	cluster = cluster.DeepCopy()
+	cluster.Status.AKSStatus.UpstreamSpec = upstreamSpec
+	return e.ClusterClient.Update(cluster)
+}
+
+// updateAKSClusterConfig updates the AKSClusterConfig object's spec with the cluster's AKSConfig if they are not equal..
+func (e *aksOperatorController) updateAKSClusterConfig(cluster *mgmtv3.Cluster, aksClusterConfigDynamic *unstructured.Unstructured, spec map[string]interface{}) (*mgmtv3.Cluster, error) {
+	list, err := e.DynamicClient.Namespace(namespace.GlobalNamespace).List(context.TODO(), v1.ListOptions{})
+	if err != nil {
+		return cluster, err
+	}
+	selector := fields.OneTermEqualSelector("metadata.name", cluster.Name)
+	w, err := e.DynamicClient.Namespace(namespace.GlobalNamespace).Watch(context.TODO(), v1.ListOptions{ResourceVersion: list.GetResourceVersion(), FieldSelector: selector.String()})
+	if err != nil {
+		return cluster, err
+	}
+	aksClusterConfigDynamic.Object["spec"] = spec
+	aksClusterConfigDynamic, err = e.DynamicClient.Namespace(namespace.GlobalNamespace).Update(context.TODO(), aksClusterConfigDynamic, v1.UpdateOptions{})
+	if err != nil {
+		return cluster, err
+	}
+
+	// AKS cluster and node pool statuses are not always immediately updated. This cause the AKSConfig to
+	// stay in "active" for a few seconds, causing the cluster to go back to "active".
+	timeout := time.NewTimer(10 * time.Second)
+	for {
+		select {
+		case event := <-w.ResultChan():
+			aksClusterConfigDynamic = event.Object.(*unstructured.Unstructured)
+			status, _ := aksClusterConfigDynamic.Object["status"].(map[string]interface{})
+			if status["phase"] == "active" {
+				continue
+			}
+
+			// this enqueue is necessary to ensure that the controller is reentered with the updating phase
+			e.ClusterEnqueueAfter(cluster.Name, enqueueTime)
+			return e.SetUnknown(cluster, apimgmtv3.ClusterConditionUpdated, "")
+		case <-timeout.C:
+			cluster, err = e.recordAppliedSpec(cluster)
+			if err != nil {
+				return cluster, err
+			}
+			return cluster, nil
+		}
+	}
+}
+
+// generateAndSetServiceAccount uses the API endpoint and CA cert to generate a service account token. The token is then copied to the cluster status.
+func (e *aksOperatorController) generateAndSetServiceAccount(cluster *mgmtv3.Cluster) (*mgmtv3.Cluster, error) {
+
+	restConfig, err := e.getKubeConfig(cluster)
+	if err != nil {
+		return cluster, err
+	}
+	saToken, err := generateSAToken(restConfig)
```

Comment:
This needs to use the `ClusterDialer` if the cluster is private, otherwise it tries to use the control plane endpoint to create the service account and can't reach it.

Reply:
I haved added support for `ClusterDialer`